trin-core/src/types/receipts.rs

```diff
@@ -0,0 +1,523 @@
+use std::ops::{Deref, DerefMut};
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use bytes::Bytes;
+use eth_trie::{EthTrie, MemoryDB, Trie};
+use ethereum_types::{Address, Bloom, BloomInput, H256, U256};
+use rlp::{self, DecoderError, Rlp, RlpStream};
+use rlp_derive::{RlpDecodable, RlpEncodable};
+
+/// Represents the `Receipts` datatype used by the chain history wire protocol
+pub struct Receipts {
+    pub receipt_list: Vec<TypedReceipt>,
+}
+
+impl Receipts {
+    pub fn root(&self) -> anyhow::Result<H256> {
+        let memdb = Arc::new(MemoryDB::new(true));
+        let mut trie = EthTrie::new(Arc::clone(&memdb));
+
+        // Insert receipts into receipts tree
+        for (index, receipt) in self.receipt_list.iter().enumerate() {
+            let path = rlp::encode(&index).freeze().to_vec();
+            let encoded_receipt = receipt.encode();
+            match trie.insert(&path, &encoded_receipt) {
+                Ok(_) => (),
+                Err(msg) => return Err(anyhow!("Error calculating receipts root: {msg:?}")),
+            }
+        }
+
+        match trie.root_hash() {
+            Ok(val) => Ok(val),
+            Err(msg) => Err(anyhow!("Error calculating receipts root: {msg:?}")),
+        }
+    }
+}
+
+// Based on https://github.com/openethereum/openethereum/blob/main/crates/ethcore/types/src/receipt.rs
+
+/// A record of execution for a `LOG` operation.
+#[derive(Default, Debug, Clone, PartialEq, Eq, RlpEncodable, RlpDecodable)]
+pub struct LogEntry {
+    /// The address of the contract executing at the point of the `LOG` operation.
+    pub address: Address,
+    /// The topics associated with the `LOG` operation.
+    pub topics: Vec<H256>,
```

Comment:
Not sure the ramification exactly, but it's represented as integers, ie~ `U256`, in python.

Reply:
Hmm, I don't have a strong argument for either type here, but since both openethereum and the [ethers-rs library](https://docs.rs/ethers/latest/ethers/types/struct.Log.html) use `H256` here, I'm inclined to follow their lead.